storage-proofs-update/src/challenges.rs

```diff
@@ -0,0 +1,141 @@
+        assert!(k < partitions);
+
+        let challenge_bit_len = sector_nodes.trailing_zeros() as usize;
+        let partition_bit_len = partitions.trailing_zeros() as usize;
+        let random_bits_per_challenge = challenge_bit_len - partition_bit_len;
+        let challenges_per_digest = Fr::CAPACITY as usize / random_bits_per_challenge;
+
+        let partition_bits = k << random_bits_per_challenge;
+
+        let challenge_count = challenge_count(sector_nodes);
+        let digests_per_partition =
+            (challenge_count as f32 / challenges_per_digest as f32).ceil() as usize;
+        let digest_index_all_partitions = k * digests_per_partition;
+
+        Challenges {
+            comm_r_new,
+            partition_bits,
+            random_bits_per_challenge,
+            challenges_per_digest,
+            digest_index_all_partitions,
+            i: 0,
+            digest_bits: Vec::with_capacity(Fr::NUM_BITS as usize),
+            challenges_remaining: challenge_count,
+        }
+    }
+}
+
+impl Iterator for Challenges {
+    // Return a `usize` (as opposed to something smaller like `u32`) because
+    // `MerkleTreeTrait::gen_proof()` takes `usize` challenges.
+    type Item = usize;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.challenges_remaining == 0 {
+            return None;
+        }
+
+        // `digest = H(comm_r_new || digest_index)` where `digest_index` is across all partitions.
+        if self.i == 0 {
+            let digest_index = Fr::from(self.digest_index_all_partitions as u64);
+            let digest: Fr =
+                <PoseidonHasher as Hasher>::Function::hash2(&self.comm_r_new, &digest_index.into())
+                    .into();
+            self.digest_bits = digest.to_le_bits().into_iter().collect();
+        }
+
+        // Derive the `i`-th challenge `c` from `digest`.
+        let c_bits = {
+            let start = self.i * self.random_bits_per_challenge;
+            let stop = start + self.random_bits_per_challenge;
+            &self.digest_bits[start..stop]
+        };
+
+        let mut c = 0;
+        for (i, bit) in c_bits.iter().enumerate() {
+            if *bit {
+                c |= 1 << i;
+            }
+        }
+        // Append the partition-index bits onto the most-significant end of `c`.
+        c |= self.partition_bits;
+
+        self.i += 1;
+        if self.i == self.challenges_per_digest {
+            self.i = 0;
+            self.digest_index_all_partitions += 1;
+        }
+        self.challenges_remaining -= 1;
+        Some(c)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use filecoin_hashers::Domain;
+    use rand::SeedableRng;
+    use rand_xorshift::XorShiftRng;
+    use storage_proofs_core::TEST_SEED;
+
+    use crate::constants::{TreeRDomain, ALLOWED_SECTOR_SIZES};
+
+    #[test]
+    fn test_challenges() {
+        let mut rng = XorShiftRng::from_seed(TEST_SEED);
```

Comment:
Given this is a seeded rng, I would suggest to generate a small list of challenges, that are fixed, check them manually and then always compare against those in tests.

Reply:
Ok, I believe that I added this in [`test_challenges_against_hardcoded`](https://github.com/filecoin-project/rust-fil-proofs/blob/ce1836f6bc5dcaef79610727e433cacf7c8bf74c/storage-proofs-update/src/challenges.rs#L127).